File: topics/CONSISTENT_HASHING/consistent_hashing.py
import hashlib
import bisect

class ConsistentHashRing:
    """
    A probabilistic-ish load balancing distribution strategy that minimizes
    reshuffling when nodes are added or removed.
    """
    def __init__(self, replicas=3):
        """
        replicas : int : The number of virtual nodes (vnodes) to create for each physical node.
                         Higher numbers provide better statistical distribution.
        """
        self.replicas = replicas
        # The Hash Ring. A sorted list of integers representing positions on the ring.
        self.ring = []
        # Maps a hash position on the ring to the physical node name.
        self._nodes = {}

    def _hash(self, key):
        """Returns a 32-bit hash of the key."""
        return int(hashlib.md5(str(key).encode('utf-8')).hexdigest(), 16)

    def add_node(self, node_name):
        """
        Adds a physical node to the ring by creating multiple virtual nodes.
        """
        for i in range(self.replicas):
            # Create a virtual node name (e.g., "ServerA_0", "ServerA_1")
            vnode_name = f"{node_name}_{i}"
            # Hash the vnode to find its position on the ring
            pos = self._hash(vnode_name)

            # Add to the ring and map to the physical node
            self.ring.append(pos)
            self._nodes[pos] = node_name

        # Keep the ring sorted so we can use binary search to walk clockwise
        self.ring.sort()

    def remove_node(self, node_name):
        """
        Removes a physical node and all its virtual nodes from the ring.
        """
        for i in range(self.replicas):
            vnode_name = f"{node_name}_{i}"
            pos = self._hash(vnode_name)

            self.ring.remove(pos)
            del self._nodes[pos]

    def get_node(self, key):
        """
        Finds the physical node responsible for the given key.
        """
        if not self.ring:
            return None

        # Hash the data key
        pos = self._hash(key)

        # Use binary search to find the first node positioned *after* the key on the ring
        # This is equivalent to "walking clockwise"
        index = bisect.bisect_right(self.ring, pos)

        # If the index is out of bounds, it means the key is placed after the last node.
        # It must wrap around to the very first node on the ring.
        if index == len(self.ring):
            index = 0

        # Return the physical node mapped to that position
        return self._nodes[self.ring[index]]

if __name__ == '__main__':
    # Sample Input: Create a ring with 3 physical servers, 100 virtual nodes each
    ring = ConsistentHashRing(replicas=100)

    servers = ["Server_A", "Server_B", "Server_C"]
    for server in servers:
        ring.add_node(server)
        print(f"Added {server} to the ring.")

    print("\nRouting Data:")
    # Route some data keys
    data_keys = ["user123", "user456", "image.png", "dataset.json"]
    for key in data_keys:
        target = ring.get_node(key)
        # Sample Output
        print(f"Key '{key}' is routed to -> {target}")

    print("\nSimulating Server_B crashing...")
    ring.remove_node("Server_B")

    print("\nRe-Routing Data:")
    for key in data_keys:
        target = ring.get_node(key)
        # Notice how most keys stay on the same server, only Server_B's keys move!
        print(f"Key '{key}' is NOW routed to -> {target}")

File: topics/CONSISTENT_HASHING/consistent_hashing.rs
use std::collections::HashMap;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

/// A Consistent Hash Ring implementation
pub struct ConsistentHashRing {
    /// Number of virtual nodes per physical node
    replicas: usize,
    /// The sorted hash ring representing node positions
    ring: Vec<u64>,
    /// Map from hash position to physical node name
    nodes: HashMap<u64, String>,
}

impl ConsistentHashRing {
    /// Creates a new hash ring
    pub fn new(replicas: usize) -> Self {
        ConsistentHashRing {
            replicas,
            ring: Vec::new(),
            nodes: HashMap::new(),
        }
    }

    /// Helper to hash any key into a u64
    fn hash<T: Hash>(&self, key: &T) -> u64 {
        let mut hasher = DefaultHasher::new();
        key.hash(&mut hasher);
        hasher.finish()
    }

    /// Adds a physical node and its virtual replicas to the ring
    pub fn add_node(&mut self, node_name: &str) {
        for i in 0..self.replicas {
            // Create a virtual node identifier
            let vnode_name = format!("{}_{}", node_name, i);
            let pos = self.hash(&vnode_name);

            self.ring.push(pos);
            self.nodes.insert(pos, node_name.to_string());
        }
        // Keep the ring sorted for binary search
        self.ring.sort_unstable();
    }

    /// Removes a physical node and its virtual replicas from the ring
    pub fn remove_node(&mut self, node_name: &str) {
        for i in 0..self.replicas {
            let vnode_name = format!("{}_{}", node_name, i);
            let pos = self.hash(&vnode_name);

            // Remove the position from the sorted ring
            if let Ok(idx) = self.ring.binary_search(&pos) {
                self.ring.remove(idx);
            }
            // Remove the mapping
            self.nodes.remove(&pos);
        }
    }

    /// Finds the node responsible for a given key
    pub fn get_node<T: Hash>(&self, key: &T) -> Option<String> {
        if self.ring.is_empty() {
            return None;
        }

        let pos = self.hash(key);

        // Binary search to find the first position strictly greater than or equal to the key's hash
        // This is walking clockwise on the ring
        let index = match self.ring.binary_search(&pos) {
            Ok(i) => i,      // Exact match (rare)
            Err(i) => i,     // Insertion point
        };

        // If the insertion point is past the end of the ring, wrap around to index 0
        let target_pos = if index == self.ring.len() {
            self.ring[0]
        } else {
            self.ring[index]
        };

        self.nodes.get(&target_pos).cloned()
    }
}

fn main() {
    // Sample Input: 3 servers, 100 replicas
    let mut ring = ConsistentHashRing::new(100);

    let servers = vec!["Server_A", "Server_B", "Server_C"];
    for server in &servers {
        ring.add_node(server);
        println!("Added {} to the ring.", server);
    }

    println!("\nRouting Data:");
    let data_keys = vec!["user123", "user456", "image.png", "dataset.json"];

    for key in &data_keys {
        if let Some(target) = ring.get_node(key) {
            println!("Key '{}' is routed to -> {}", key, target);
        }
    }

    println!("\nSimulating Server_B crashing...");
    ring.remove_node("Server_B");

    println!("\nRe-Routing Data:");
    for key in &data_keys {
        if let Some(target) = ring.get_node(key) {
            println!("Key '{}' is NOW routed to -> {}", key, target);
        }
    }
}
